From c92a49ad7b6cccace653264e9711d4b1d659198c Mon Sep 17 00:00:00 2001
From: Vitaly Kuznetsov <vkuznets@redhat.com>
Date: Wed, 7 Jan 2015 17:58:04 -0500
Subject: [scsi] storvsc: Fix a bug in storvsc limits

Message-id: <1420653484-6399-1-git-send-email-vkuznets@redhat.com>
Patchwork-id: 102522
O-Subject: [RHEL7.1 KERNEL PATCH] Drivers: scsi: storvsc: Fix a bug in storvsc limits
Bugzilla: 1174162
RH-Acked-by: Radim Krcmar <rkrcmar@redhat.com>
RH-Acked-by: Paolo Bonzini <pbonzini@redhat.com>
RH-Acked-by: Jason Wang <jasowang@redhat.com>
RH-Acked-by: Amos Kong <akong@redhat.com>

Bugzilla: https://bugzilla.redhat.com/show_bug.cgi?id=1174162
Brew: https://brewweb.devel.redhat.com/taskinfo?taskID=8467629
Tested: Win2008R2, Gen1 and Gen2 guests with Win2012R2 by me

This is a regression compared to RHEL7.0 and can potentially lead to issues.

Upstream commit 4ca85db910add31139e417e327be84ced2652a10
 (scsi/misc branch of git://git.kernel.org/pub/scm/linux/kernel/git/jejb/scsi.git)
Author: K. Y. Srinivasan <kys@microsoft.com>
Date:   Tue Dec 16 13:21:44 2014 -0800

    Commit 4cd83ecdac20d30725b4f96e5d7814a1e290bc7e changed the limits to
    reflect the values on the host. It turns out that WS2008R2 cannot
    correctly handle these new limits. Fix this bug by setting the limits
    based on the host.

    Signed-off-by: K. Y. Srinivasan <kys@microsoft.com>

Signed-off-by: Vitaly Kuznetsov <vkuznets@redhat.com>
Signed-off-by: Jarod Wilson <jarod@redhat.com>

diff --git a/drivers/scsi/storvsc_drv.c b/drivers/scsi/storvsc_drv.c
index c2c6515..ea15943 100644
--- a/drivers/scsi/storvsc_drv.c
+++ b/drivers/scsi/storvsc_drv.c
@@ -1760,6 +1760,9 @@ static int storvsc_probe(struct hv_device *device,
  bool dev_is_ide = ((dev_id->driver_data == IDE_GUID) ? true : false);
  int target = 0;
  struct storvsc_device *stor_device;
+ int max_luns_per_target;
+ int max_targets;
+ int max_channels;
 
  /*
   * Based on the windows host we are running on,
@@ -1773,12 +1776,18 @@ static int storvsc_probe(struct hv_device *device,
   vmscsi_size_delta = sizeof(struct vmscsi_win8_extension);
   vmstor_current_major = VMSTOR_WIN7_MAJOR;
   vmstor_current_minor = VMSTOR_WIN7_MINOR;
+  max_luns_per_target = STORVSC_IDE_MAX_LUNS_PER_TARGET;
+  max_targets = STORVSC_IDE_MAX_TARGETS;
+  max_channels = STORVSC_IDE_MAX_CHANNELS;
   break;
  default:
   sense_buffer_size = POST_WIN7_STORVSC_SENSE_BUFFER_SIZE;
   vmscsi_size_delta = 0;
   vmstor_current_major = VMSTOR_WIN8_MAJOR;
   vmstor_current_minor = VMSTOR_WIN8_MINOR;
+  max_luns_per_target = STORVSC_MAX_LUNS_PER_TARGET;
+  max_targets = STORVSC_MAX_TARGETS;
+  max_channels = STORVSC_MAX_CHANNELS;
   break;
  }
 
@@ -1826,9 +1835,9 @@ static int storvsc_probe(struct hv_device *device,
   break;
 
  case SCSI_GUID:
-  host->max_lun = STORVSC_MAX_LUNS_PER_TARGET;
-  host->max_id = STORVSC_MAX_TARGETS;
-  host->max_channel = STORVSC_MAX_CHANNELS - 1;
+  host->max_lun = max_luns_per_target;
+  host->max_id = max_targets;
+  host->max_channel = max_channels - 1;
   break;
 
  default:
-- 
1.7.1